### ISSUE TITLE
lovely waste of everyone's time to keep repeating a wide known inability to track source IP

### ISSUE THREAD

**Issue by jorhett** [kubernetes-sigs/gateway-api#2609]
**What would you like to be added**:

Someone in the Gateway SIG who prioritizes solving problems, rather than prioritizing insulting and attacking people, then arbitrarily closing their tickets without any reasonable justification.

**Why this is needed**:

So we've posted #451 which was closed not because the problem was solved, but because you just really wanted a fresh issue. So I wasted my time to open a new issue #2596 restating the exact same problem, for which you closed it again saying you wanted one more piece of information (which existed already in the original issue 3 times). 

To quote @robscott about how everyone is supposed to behave:
> Demonstrating empathy and kindness toward other people
> Being respectful of differing opinions, viewpoints, and experiences
> I'd also mention that many of us are spending personal time to work on this project. 

@robscott and @youngnick closed these tickets just to show how powerful they are, how they don't have to care, don't have to show empathy or kindness, and have no problem wasting our time. The brazenly disrespectful attitude shown makes it clear that their time is valuable, and yet ours is to be wasted as often as possible. And to demonstrate that they are above the code of conduct, and can be as rude and unhelpful as they please, and can waste other's time to their heart's content.

Every bit of information that @youngnick asked for in new issue #2596 was already present in #451. Now @robscott requested that yet another issue be created will the exact same information available in #2596. Same problem, same need, no more information than was included in either #451 or #2596 but you want to make me run around and repeat my efforts just because you can force me to, as if I have nothing better to do. I could be writing code for a PR to implement this but no, it's much better if you get to take a piss on me. This is not appreciated. You need to **go read your own code of conduct** and **consider what it means**.

> open an issue that describes the features you actually want to be configurable. ExternalTrafficPolicy=Local is really two things:

I only mentioned one of those things, and I mentioned it three times in the issue. There's no lack of clarity there, nor was there lack of clarity in #451 either.

> Although you're correct that many implementations currently create a Service of type LB as part of their implementation of a Gateway, that is an implementation detail, it is not actually part of Gateway API. For example, some implementations 

"Some implementations" (by which I assume you mean _other_) was never, not even once mentioned in my issue report. I reported that the Gateway API as released by this SIG creates a service when a gateway is created. I used `kubectl get crd gateways.gateway.networking.k8s.io` which is YOUR release.  There is no other implementation to blame, no other group involved in how you implemented this. Your design, your code, your release.

You can build a fresh cluster with [metallb](https://github.com/metallb/metallb) and your APIs and nothing else and recreate this problem. It's well described in the metallb docs. It's describe in the Istio docs. It's widely reported on stack overflow. It's talked about in your Slack channels. There is no ambiguity here. There never was, and pretending it's unclear as an excuse to CLOSE the issue and waste everyone's time yet again is utterly ridiculous.

Yes, I'm sure you felt that your time was wasted reading this issue: but that decision was made by @robscott not by me. If he had requested that I add some more details to the previous issue, I would have gladly done so. Instead he wanted to demonstrate that he's above and impervious to your code of conduct, and willing to waste all of our time ad nauseum.

## How can we fix this?

1. Apply your code of conduct to your maintainers

2. Re-open #451 or #2596 and ask your questions. It's totally fair to ask for more information, or to ask me to rephrase something to be more respectful. It's not reasonable to close the issue because it's missing a piece of information, and you'll like to watch others waste their time repeating themselves over and over and over again. It's not reasonable to complain about respect, after making someone run in useless, irrelevant circles to copypasta the exact same information to you over and over and over again.

I'm doing your work for you--in that this issue is broadly known, impacting many projects, and yet you wouldn't work on it without someone wasting their time to repeat it again for you YEARS after the original report. So I wasted my time to recreate it, and you acted with empathy and respect for me having wasted my time doing copypasta work for you?  Hahahaha yeah.

I probably would've created the implementation too. But now that you've gone out of your way to take a piss on me twice, and I'm sure you'll repeat yourself and do it again because solving problems clearly isn't of interest; taking a piss on people who would help is apparently much more important than working code.

TL;DR

* You have a well-known, well-documented in #196 problem that people have been asking you for years to address. A FUNDAMENTAL PROBLEM YOUR OWN TEAM OUTLINED AS REQUIRED IN YOUR ORIGINAL PROPOSAL BUT DIDN'T SOLVE. You knew it 4 years ago. You had open issues dating back years, and rather than fix them you wanted someone to waste their time repeating the same information again. 

* If you have a problem getting technical expertise to assist you, you can look at #451 and #2596 and see exactly why. When you attack and insult others, you create frustration. And yet @robscott went out of his way to repeat it again nearly instantly, apparently because he saw @youngnick take a step back and offer some respect. No way was @robscott going to stand for that.

* I was fully intending to build the solution myself as soon as you confirmed the approach. But I'm busy too, far too busy to keep wasting my time running in circles repeating the same issue over and over and over again. Or to be your little spin-toy to waste my time.

I'm absolutely certain you'll take an shot at me, or ban me for pointing out how badly you treat people. Go for it. No wonder you don't have enough contributors, you've made it painfully clear that they exist only to run like ants when you stomp your foot.

**Comment by Xunzhuo** [comment]
Hey @jorhett 

Thank you once again for your feedback. Our community should be collaborated on friendliness and open communication, not too much opposition and conflict. We are committed to fostering an environment that encourages collaboration, understanding, and mutual respect. Such an atmosphere not only promotes more effective communication but also sparks greater innovation and cooperation. I believe we are doing this not that bad, which is one of the reasons that the Gateway API has attracted the participation of many API Gateway implementations / contributors (including me).

I would like to emphasize the significant contributions of gwapi maintainer team to the project. Their work is crucial to the project's success. We are deeply grateful for their hard work and the friendly attitude they bring to our community. We understand that without their relentless efforts, our project would not have achieved its current success. 

I offen talked with @robscott, @youngnick (Previous Envoy Gateway Maintainer), @shaneutt  in private, I can say that they forks are really nice and kind, and they do care about everyone`s idea, when discussing it clearly in peace and respect.

The Gateway API project is the result of the collective effort of many implementions, and we recognize that each change requires careful consideration to ensure its applicability across different implementation scenarios. GWAPI team does indeed need a clear understanding of specific use cases when evaluating proposals. This not only helps us assess whether new modifications are suitable for all implementations but also ensures that our decisions meet the broad needs of the community.

Therefore, when you suggest adding a specific field, we would prefer to understand its application in a common specific scenarios. Such detailed descriptions will help us and all implementions better understand its necessity and evaluate its applicability to each implementation. We encourage you to share more context and use cases, which will not only help us gain a deeper understanding of the issue but also facilitate our collaboration in finding the most appropriate solution.

Crafting APIs is a challenging task, something I've experienced firsthand while maintaining the Envoy Gateway. We cannot endlessly add new fields to an API definition every time a user mentions a new scenario. We must always evaluate whether a scenario is common enough, because if an API field is not widely applicable, it means there will be significant maintainability issues in the future, which is a nightmare for maintainers. So, the reality is that creating a universal API is the result of collaboration and discussion among multiple parties. So please discuss it in peace and in patience, and if most of implementations request for that feature, I am sure gwapi team will accept that proposal naturally.

BTW, GWAPI is flexible and provides many extension points like PolicyAttachment for Gateway / xRoute and parametersRefs in GatewayClass. In Envoy Gateway, we have EnvoyProxy CRD refered by GatewayClass parametersRefs which is used to manage the EnvoyProxy kubernetes infra, including serviceType, and this kind of mechanism also can resolve your case.

Thank you once again for your valuable time.

**Comment by jorhett** [comment]
> they do care about everyone`s idea, when discussing it clearly in peace and respect.

Closing an issue prevents discussion, so I haven't been able to witness this yet.

> Working on APIs is a challenging task

Absolutely. I expected that finding the right implementation may take some time, the discussion might evolve, the end solution might look very different from when I propose. I'm totally open to that. It's very difficult to have any sort of discussion when after a single post the issue is closed with instructions to open a new one. Discussion across hundreds of separate issues is hard to track for everyone involved.

> GWAPI is flexible and provides many extension points like PolicyAttachment for Gateway / xRoute and parametersRefs in GatewayClass. In Envoy Gateway, we have EnvoyProxy CRD refered by GatewayClass parametersRefs which is used to manage the EnvoyProxy kubernetes infra, including service/deployment etc, and this kind of mechanism also can resolve your case.

I'm absolutely open to any solution which solves the underlying requirement. Direct access to the authoritative source IP, without having to (attempt to) pull that out of a long string which can be mangled or repeated in the header, and would require every implementation to implement that logic exactly the same (implausible -> impossible)

**Comment by robscott** [comment]
Hey @jorhett, we really do want to have a respectful discussion here, we just need to keep the discussion focused on the substance and avoid any attacks on individuals.

>Your code. Your releases create Services. There is a problem with how they are created. Absolutely nothing in this paragraph is relevant or topical to the problem reported. This was insulting and dismissive.

I think this is an important misunderstanding here. To be clear, the scope of Gateway API is:

* API spec and validation
* Conformance tests
* Documentation

If you install these CRDs without a corresponding [implementation](https://gateway-api.sigs.k8s.io/implementations/), they're going to be effectively useless. Nothing in this repo creates a Service to implement a Gateway. Instead, what you're observing is a common (but not universal) implementation detail. For many implementations, a Gateway is effectively implemented by pointing a Service type=LB to a set of dataplane Pods. 

Note that this is just one possible implementation. Implementations from cloud providers (GKE, EKS, AKS) translate Gateway API directly to cloud load balancing resources. I believe it's also possible to deploy other implementations outside of a cluster entirely (and by extension, also without a Service type=LB). 

Although it is admittedly very common for a Service type=LB to be used to implement a Gateway today, that is not something we want to be stuck with long term. As you're likely aware, we're also trying to replace Service type=LB with L4 Gateways. Ideally in the future, implementations that rely on Cloud L4 LBs would be able to configure them via Gateway API instead of relying on Service type=LB. 

All of that background may help explain why we're trying to redirect from configuring specific fields on Service to instead focus on higher level ideas and feature requests. We're trying to find sufficiently generic ways to describe a feature that they can be meaningful whether or not a LB Service is used by the underlying implementation to provision a Gateway.

I also appreciate @Xunzhuo's kind words here. He pointed to how Envoy Gateway allows users to customize details of the resources it uses to implement a Gateway (Service and Deployment) via [GatewayClass Params](https://gateway.envoyproxy.io/v0.6.0/user/customize-envoyproxy/). This is the path we'd recommend for implementations that want to support configuring something like ExternalTrafficPolicy today. 

Of course if we can define the feature in a more generic way that can be useful beyond just implementations that use LB Services, it could be helpful to have a portable way to describe this directly within Gateway API.
